src/Language/PureScript/TypeChecker/Types.hs

```diff
@@ -173,12 +174,29 @@ typesOf bindingGroupType moduleName vals = withFreshSubstitution $ do
     constrain cs ty = foldr srcConstrainedType ty (map (\(_, _, x) -> x) cs)
 
     -- Apply the substitution that was returned from runUnify to both types and (type-annotated) values
-    tidyUp ts sub = first (map (second (first (second (overTypes (substituteType sub) *** substituteType sub))))) ts
+    tidyUp ts sub = first (map (second (first (second (updateExpr sub *** substituteType sub))))) ts
```

Comment:
Should the fix not be to modify `overTypes`?

Reply:
I guess I'd been a bit reluctant to modify `overTypes` since it's used the once elsewhere. But the change looks like it'd be consistent with that other use site, and it passes tests, so your suggestion would the clearer option.

Anyway, before I push new code, what I think you have in mind follows. But let me know if you have any other ideas/recommendations.

```purescript
-- Types.hs
-- DELETE `updateExpr`, `updateContextsInExpr`, and `updateTypeAnnotations`.
tidyUp ts sub = first (map (second (first (second (overTypes (substituteType sub) *** substituteType sub))))) ts
```
```purescript
-- Traversals.purs
overTypes :: (SourceType -> SourceType) -> Expr -> Expr
overTypes f = let (_, f', _) = everywhereOnValues id g id in f'
  where
  g :: Expr -> Expr
  g (TypedValue checkTy val t) = TypedValue checkTy val (f t)
  g (TypeClassDictionary c sco hints) =
    TypeClassDictionary
      (mapConstraintArgs (fmap f) c)
      (updateCtx sco)
      hints
  g other = other
  updateDict f dict = dict { tcdInstanceTypes = f (tcdInstanceTypes dict) }
  updateScope = fmap . fmap . fmap . fmap $ updateDict $ fmap f
  updateCtx = M.alter updateScope Nothing
```

(By the way, I apologize for my delayed response. I hadn't even noticed your comment until I just happened by this pull request yesterday. Also, thanks; I appreciate your input.)